Allow disabling "g" prefix for commit hash in version number

I've had multiple folks (including myself) be confused about the `g` in front of the commit sha of a version number like `16.0.40-preview+ge6c3a1f9e4`.

You don't immediately realize without knowing this convention that the `g` isn't part of the sha and wonder why you can't find the commit. Is there a reason for the prefix or could there be an option to turn it off?

So... now I know why the `g` prefix is there. Without it, if the commit id happens to start with a number, nuget produces warnings:

This works great (when the commit starts with a letter):
```
Successfully created package '/root/bin/Debug/root.42.42.4-f92b4f1cbd.nupkg'.
```

But when the commit starts with a number:
```
Successfully created package '/root/bin/Debug/root.42.42.3-7d637a3dc6.nupkg'.
/usr/share/dotnet/sdk/2.1.403/Sdks/NuGet.Build.Tasks.Pack/build/NuGet.Build.Tasks.Pack.targets(199,5): warning NU5105: The package version '42.42.3-7d637a3dc6' uses SemVer 2.0.0 or components of SemVer 1.0.0 that are not supported on legacy clients. Change the package version to a SemVer 1.0.0 string. If the version contains a release label it must start with a letter. This message can be ignored if the package is not intended for older clients. [/root/root.csproj]
```

So, that stinks. 

I *could* still make it an option, or make the `g` appear in only certain contexts and not others, but I'm worried it will be confusing/frustrating if the `g` appears in some places but not others since folks will have an even harder time predicting and understanding when/where it appears.

--- a/src/NerdBank.GitVersioning.Tests/BuildIntegrationTests.cs
+++ b/src/NerdBank.GitVersioning.Tests/BuildIntegrationTests.cs
@@ -176,7 +176,7 @@ public class BuildIntegrationTests : RepoTestBase
         Assumes.True(repo.Index[VersionFile.JsonFileName] == null);
         var buildResult = await this.BuildAsync();
         Assert.Equal("3.4.0." + repo.Head.Commits.First().GetIdAsVersion().Revision, buildResult.BuildVersion);
-        Assert.Equal("3.4.0+g" + repo.Head.Commits.First().Id.Sha.Substring(0, 10), buildResult.AssemblyInformationalVersion);
+        Assert.Equal("3.4.0+" + repo.Head.Commits.First().Id.Sha.Substring(0, 10), buildResult.AssemblyInformationalVersion);
     }
 
     [Fact]
@@ -201,7 +201,7 @@ public class BuildIntegrationTests : RepoTestBase
         repo.Commit("empty", this.Signer, this.Signer, new CommitOptions { AllowEmptyCommit = true });
         var buildResult = await this.BuildAsync();
         Assert.Equal("0.0.1." + repo.Head.Commits.First().GetIdAsVersion().Revision, buildResult.BuildVersion);
-        Assert.Equal("0.0.1+g" + repo.Head.Commits.First().Id.Sha.Substring(0, 10), buildResult.AssemblyInformationalVersion);
+        Assert.Equal("0.0.1+" + repo.Head.Commits.First().Id.Sha.Substring(0, 10), buildResult.AssemblyInformationalVersion);
     }
 
     [Fact]
@@ -915,7 +915,7 @@ public class BuildIntegrationTests : RepoTestBase
         Version assemblyVersion = GetExpectedAssemblyVersion(versionOptions, version);
         var additionalBuildMetadata = from item in buildResult.BuildResult.ProjectStateAfterBuild.GetItems("BuildMetadata")
                                       select item.EvaluatedInclude;
-        var expectedBuildMetadata = $"+g{commitIdShort}";
+        var expectedBuildMetadata = $"+{commitIdShort}";
         if (additionalBuildMetadata.Any())
         {
             expectedBuildMetadata += "." + string.Join(".", additionalBuildMetadata);
@@ -944,7 +944,8 @@ public class BuildIntegrationTests : RepoTestBase
 
         // NuGet is now SemVer 2.0 and will pass additional build metadata if provided
         bool semVer2 = versionOptions?.NuGetPackageVersionOrDefault.SemVer == 2;
-        string pkgVersionSuffix = buildResult.PublicRelease ? string.Empty : $"-g{commitIdShort}";
+        string semVer1CommitPrefix = semVer2 ? string.Empty : "g";
+        string pkgVersionSuffix = buildResult.PublicRelease ? string.Empty : $"-{semVer1CommitPrefix}{commitIdShort}";
         if (semVer2)
         {
             pkgVersionSuffix += expectedBuildMetadataWithoutCommitId;
@@ -965,7 +966,7 @@ public class BuildIntegrationTests : RepoTestBase
             Assert.Equal(expectedVersion, buildNumberSemVer.Version);
             Assert.Equal(buildResult.PrereleaseVersion, buildNumberSemVer.Prerelease);
             string expectedBuildNumberMetadata = hasCommitData && commitIdOptions.WhereOrDefault == VersionOptions.CloudBuildNumberCommitWhere.BuildMetadata
-                ? $"+g{commitIdShort}"
+                ? $"+{commitIdShort}"
                 : string.Empty;
             if (additionalBuildMetadata.Any())
             {

--- a/src/NerdBank.GitVersioning.Tests/VersionOracleTests.cs
+++ b/src/NerdBank.GitVersioning.Tests/VersionOracleTests.cs
@@ -240,7 +240,7 @@ public class VersionOracleTests : RepoTestBase
         this.InitializeSourceControl();
         var oracle = VersionOracle.Create(this.RepoPath);
         oracle.PublicRelease = false;
-        Assert.Equal($"7.8.9-foo.25.g{this.CommitIdShort}", oracle.NuGetPackageVersion);
+        Assert.Equal($"7.8.9-foo.25.{this.CommitIdShort}", oracle.NuGetPackageVersion);
     }
 
     [Fact]

--- a/src/NerdBank.GitVersioning/VersionOracle.cs
+++ b/src/NerdBank.GitVersioning/VersionOracle.cs
@@ -143,7 +143,7 @@
             {
                 if (!string.IsNullOrEmpty(this.GitCommitId))
                 {
-                    yield return $"g{this.GitCommitId.Substring(0, 10)}";
+                    yield return this.GitCommitId.Substring(0, 10);
                 }
 
                 foreach (string identifier in this.BuildMetadata)
@@ -335,7 +335,7 @@
         /// <summary>
         /// Gets the version to use for NuGet packages.
         /// </summary>
-        public string NuGetPackageVersion => this.VersionOptions?.NuGetPackageVersionOrDefault.SemVerOrDefault == 1 ? this.SemVer1 : this.SemVer2;
+        public string NuGetPackageVersion => this.VersionOptions?.NuGetPackageVersionOrDefault.SemVerOrDefault == 1 ? this.NuGetSemVer1 : this.SemVer2;
 
         /// <summary>
         /// Gets the version to use for NPM packages.
@@ -343,14 +343,14 @@
         public string NpmPackageVersion => this.SemVer2;
 
         /// <summary>
-        /// Gets a SemVer 1.0 compliant string that represents this version, including the -gCOMMITID suffix
+        /// Gets a SemVer 1.0 compliant string that represents this version, including the -COMMITID suffix
         /// when <see cref="PublicRelease"/> is <c>false</c>.
         /// </summary>
         public string SemVer1 =>
             $"{this.Version.ToStringSafe(3)}{this.PrereleaseVersionSemVer1}{this.SemVer1BuildMetadata}";
 
         /// <summary>
-        /// Gets a SemVer 2.0 compliant string that represents this version, including a +gCOMMITID suffix
+        /// Gets a SemVer 2.0 compliant string that represents this version, including a +COMMITID suffix
         /// when <see cref="PublicRelease"/> is <c>false</c>.
         /// </summary>
         public string SemVer2 =>
@@ -361,8 +361,30 @@
         /// </summary>
         public int SemVer1NumericIdentifierPadding => this.VersionOptions?.SemVer1NumericIdentifierPaddingOrDefault ?? 4;
 
-        private string SemVer1BuildMetadata =>
+        /// <summary>
+        /// Gets the build metadata, compliant to the NuGet-compatible subset of SemVer 1.0.
+        /// </summary>
+        /// <remarks>
+        /// When adding the git commit ID in a -prerelease tag, prefix a `g` because
+        /// older NuGet clients (the ones that support only a subset of semver 1.0)
+        /// cannot handle prerelease tags that begin with a number (which a git commit ID might).
+        /// See <see href="https://github.com/AArnott/Nerdbank.GitVersioning/issues/260#issuecomment-445511898">this discussion</see>.
+        /// </remarks>
+        private string NuGetSemVer1BuildMetadata =>
             this.PublicRelease ? string.Empty : $"-g{this.GitCommitIdShort}";
+
+        /// <summary>
+        /// Gets the build metadata, compliant to SemVer 1.0.
+        /// </summary>
+        private string SemVer1BuildMetadata =>
+            this.PublicRelease ? string.Empty : this.GitCommitIdShort;
+
+        /// <summary>
+        /// Gets a SemVer 1.0 compliant string that represents this version, including the -gCOMMITID suffix
+        /// when <see cref="PublicRelease"/> is <c>false</c>.
+        /// </summary>
+        private string NuGetSemVer1 =>
+            $"{this.Version.ToStringSafe(3)}{this.PrereleaseVersionSemVer1}{this.NuGetSemVer1BuildMetadata}";
 
         /// <summary>
         /// Gets the build metadata that is appropriate for SemVer2 use.
@@ -377,7 +399,7 @@
 
         private string PrereleaseVersionSemVer1 => MakePrereleaseSemVer1Compliant(this.PrereleaseVersion, this.SemVer1NumericIdentifierPadding);
 
-        private string GitCommitIdShortForNonPublicPrereleaseTag => (string.IsNullOrEmpty(this.PrereleaseVersion) ? "-" : ".") + $"g{this.GitCommitIdShort}";
+        private string GitCommitIdShortForNonPublicPrereleaseTag => (string.IsNullOrEmpty(this.PrereleaseVersion) ? "-" : ".") + this.GitCommitIdShort;
 
         private VersionOptions.CloudBuildNumberOptions CloudBuildNumberOptions { get; }
 

